plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.oracle;
+
+import io.trino.plugin.jdbc.JdbcTypeHandle;
+import io.trino.plugin.jdbc.expression.AbstractRewriteCast;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VarcharType;
+import oracle.jdbc.OracleTypes;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.trino.plugin.oracle.OracleClient.ORACLE_CHAR_MAX_CHARS;
+import static io.trino.plugin.oracle.OracleClient.ORACLE_VARCHAR2_MAX_CHARS;
+
+public class RewriteCast
+        extends AbstractRewriteCast
+{
+    public RewriteCast(BiFunction<ConnectorSession, Type, String> toTargetType)
+    {
+        super(toTargetType);
+    }
+
+    @Override
+    protected String buildCast(Type sourceType, Type targetType, String expression, String castType)
+    {
+        if (sourceType instanceof CharType sourceCharType && targetType instanceof CharType targetCharType) {
+            if (sourceCharType.getLength() < targetCharType.getLength()) {
+                // Do not cast unnecessary with extra space padding when target char type has more length than source char type
+                return expression;
```

Comment:
Here are the execution plans. When we skip rewriting `ScanProject` node gets populated and when we perform pushdown `TableScan` node is populated. So In terms of the test case, we can not use `isFullyPushedDown` when we skip rewriting.

When we skip rewriting,

```
MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [expr]
    Output partitioning: SINGLE []
    Output[columnNames = [_col0]]
    │   Layout: [expr:char(32)]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    │   _col0 := expr
    └─ ScanProject[table = oracle:presto_test_user.left_table_se623gecjf PRESTO_TEST_USER.LEFT_TABLE_SE623GECJF columns=[C_CHAR_10:char(10):CHAR]]
           Layout: [expr:char(32)]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}/{rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           expr := CAST(c_char AS char(32))
           c_char := C_CHAR_10:char(10):CHAR

]], types=[varchar(650)], setSessionProperties={}, resetSessionProperties=[]}
```

And we perform pushdown,

```
MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [pfgnrtd]
    Output partitioning: SINGLE []
    Output[columnNames = [_col0]]
    │   Layout: [pfgnrtd:char(32)]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    │   _col0 := pfgnrtd
    └─ TableScan[table = oracle:Query[SELECT "C_CHAR_10" AS "pfgnrtd_0" FROM "PRESTO_TEST_USER"."LEFT_TABLE_BJL0LGPNFS"] columns=[pfgnrtd_0:char(32):char]]
           Layout: [pfgnrtd:char(32)]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           pfgnrtd := pfgnrtd_0:char(32):char

]], types=[varchar(581)], setSessionProperties={}, resetSessionProperties=[]}
```

Reply:
rewriting has a advantage that we could use `isFullyPushedDown` to verify the pushdown in such cases.

Reply:
Also not rewriting cast for this will result in `JOIN` not pushdown for `SELECT l.id FROM table1 l JOIN table2 r ON CAST(l.c_char_10 AS char(50)) = r.c_char_50;`.

```
Output[columnNames = [id]]
│   Layout: [id:decimal(10,0)]
└─ InnerJoin[criteria = (expr = c_char_26), distribution = PARTITIONED]
   │   Layout: [id:decimal(10,0)]
   │   Distribution: PARTITIONED
   ├─ RemoteExchange[type = REPARTITION]
   │  │   Layout: [id:decimal(10,0), expr:char(50)]
   │  └─ ScanProject[table = oracle:presto_test_user.left_table_wf7cu8lky4 PRESTO_TEST_USER.LEFT_TABLE_WF7CU8LKY4 columns=[ID:decimal(10,0):NUMBER, C_CHAR_10:char(10):CHAR]]
   │         Layout: [id:decimal(10,0), expr:char(50)]
   │         expr := CAST(c_char AS char(50))
   │         c_char := C_CHAR_10:char(10):CHAR
   │         id := ID:decimal(10,0):NUMBER
   └─ LocalExchange[partitioning = HASH, arguments = [c_char_26::char(50)]]
      │   Layout: [c_char_26:char(50)]
      └─ RemoteExchange[type = REPARTITION]
         │   Layout: [c_char_26:char(50)]
         └─ TableScan[table = oracle:presto_test_user.right_table_g1rzz51027 PRESTO_TEST_USER.RIGHT_TABLE_G1RZZ51027 columns=[C_CHAR_50:char(50):CHAR]]
                Layout: [c_char_26:char(50)]
                c_char_26 := C_CHAR_50:char(50):CHAR
```